flybywiresim/a32nx
# fix(mcdu): Fix input issues on PERF/W&B and fuel prediction pages

Fixes #3525
Fixes #3321
Fixes #3528

## Summary of Changes

Fixed a number of missing or incorrect unit conversions relating to fuel and payload weight conversions when using imperial units. This fixes a number of issues where entered values were either not stored correctly or not displayed correctly. The following fields are fixed:

- PERF/W&B
   -  Page 1
      - BLOCK FUEL
      - TAXI FUEL
      - TRIP FUEL
   - Page 2
     - PAYLOAD
     - FWD BAG
     - REAR BAG
- Fuel prediction pages (INIT page B and the FUEL PRED page)
   - TAXI (on INIT B)
   - RTE RSV/%
   - ALTN/TIME
   - FINAL/TIME
   - MIN DEST FOB
   - EXTRA/TIME

Max allowed block fuel value is also changed to 19046 kg to match MSFS fuel dialog and actual fuel capacity.
Also fixes an issue where TRIP FUEL input on the PERF/W&B page was incorrectly bound to 2R.
## Screenshots (if necessary)


Before             |  After
:-------------------------:|:-------------------------:
![fuel_conversion_bug](https://user-images.githubusercontent.com/5500848/107995742-17a4d280-6fdf-11eb-9e5b-1e8628651d2d.gif) | ![fuel_conversion_fixed](https://user-images.githubusercontent.com/5500848/107995749-1a9fc300-6fdf-11eb-9aa0-c09e548fa998.gif)
![payload_conversion_bug](https://user-images.githubusercontent.com/5500848/107993418-ae22c500-6fda-11eb-9714-6d0f9b7d0c5a.gif) | ![payload_conversion_fixed](https://user-images.githubusercontent.com/5500848/107993048-53896900-6fda-11eb-99cd-3697341d477d.gif)
![fuel_pred_bug](https://user-images.githubusercontent.com/5500848/107994633-55a0f700-6fdd-11eb-84e3-ce8c6f20ba54.gif) | ![fuel_pred_fixed](https://user-images.githubusercontent.com/5500848/107995008-1d4de880-6fde-11eb-9345-50571e47b762.gif)

## References


Max fuel capacity from sources in #2286, adjusted to 19046 kg to account for rounding in MSFS




## Additional context



Discord username (if different from GitHub): Jethro#4431

## Testing instructions


Test this PR with both imperial and metric units (MCDU MENU -> OPTIONS -> FMGC -> WEIGHT UNIT, make sure to reload the aircraft after changing units): 
1. Make sure that fuel and payload values on PERF/W&B (MCDU MENU -> ATSU -> AOC MENU -> PERF/W&B) are entered correctly and that the correct amount of fuel is loaded when requesting refueling
2. Enter FROM/TO and ALTN on INIT page A 
3. Verify that both fuel prediction pages (INIT page B and FUEL PRED) behave as expected when changing field values

Note that this PR is focused around input validation and unit conversions (with some corresponding improvements to the presentation on the fuel prediction pages). Please bear this in mind when reporting potential issues with calculated values or any underlying algorithms. Also note that I have attempted to make minimal changes to the PERF/W&B payload page since this will be overhauled in an upcoming PR.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link
